### ISSUE TITLE
Better Unicode handling for Windows

### ISSUE THREAD

**Issue by brianlheim** [erikd/libsndfile#393]
Hello!

I'm a maintainer on the [SuperCollider](https://github.com/supercollider/supercollider) project, which has used libsndfile for a long time. Recently we've been doing some work to make sure that the application correctly handles Unicode path names, and it's been a little difficult to use this library's API that way (see https://github.com/supercollider/supercollider/pull/3720). The problem is that there isn't an easy way to pass a Unicode filename to the library on both Windows and Unix platforms. Essentially, I've had to write wrappers around libsndfile functions to ensure the right cross-platform behavior. Anyone who wants to use this library and get the correct behavior would have to do the same.

Since the operations that currently take `std::string` and `char const*` are broken for Windows usage anyway, it seems like it should be totally acceptable for those functions to treat input as UTF-8 and change encodings under the hood. That way, cross-platform code can just call those functions and get good behavior for free.

I'd recommend this advice from [UTF-8 Everywhere](http://utf8everywhere.org/):

- perform widening/narrowing operations as close as possible to API calls
- encode everything as UTF-8 where possible
- never store wchar/wstring data
- `char const*` and `std::string` are considered UTF-8 everywhere in the program

I'd be happy to work on a patch for this if you agree.

**Comment by evpobr** [comment]
Funny, `sf_wchar_open()` converts `wpath` to UTF-8, not to ANSI codepage for logging purposes.

**Comment by evpobr** [comment]
@erikd, revert back that PR #394.

We pass all the tests, but i tried to open sound file with cyrillic letters, just to be sure everything is OK. It was failed to open, `wpath` contains garbage.

> Since the operations that currently take std::string and char const* are broken for Windows usage anyway, it seems like it should be totally acceptable for those functions to treat input as UTF-8 and change encodings under the hood. That way, cross-platform code can just call those functions and get good behavior for free.

We canot treat input as UTF-8 aware. Only first 128 characters of ASCII are compatible with UTF-8 and any ASCII-based encoding, national encodings use character codes above. In this case UTF-8 encoded path will be broken.

**Comment by erikd** [comment]
Before we try something like this again, i suspect we may need an extra test :)

**Comment by brianlheim** [comment]
> We pass all the tests, but i tried to open sound file with cyrillic letters, just to be sure everything is OK. It was failed to open, wpath contains garbage.

I'm not sure how you tried to open the sound file, but perhaps the filename you were passing into `sf_open` was in an encoding other than UTF-8. Looking at https://github.com/erikd/libsndfile/commit/0dbaebb6a3c219a63b8e8ea20865005cb9c57910, I see a problem - you call `sf_wchar_open` if `MultiByteToWideChar` returns `0`, but according to [MSDN](https://msdn.microsoft.com/en-us/library/windows/desktop/dd319072(v=vs.85).aspx), a return of `0` indicates failure.

> We canot treat input as UTF-8 aware. Only first 128 characters of ASCII are compatible with UTF-8 and any ASCII-based encoding, national encodings use character codes above. In this case UTF-8 encoded path will be broken.

I'm not sure I follow... you absolutely can treat input as UTF-8 encoded, this just means that the burden is on the caller to ensure their data is in the correct encoding. As long as that's clearly documented, there should be no issue.

**Comment by Ghabry** [comment]
Alternatively you can use sf_open_fd (use fileno(f) on your _wfopen'ed FILE*) or write a very thin VIO wrapper, e.g.: https://github.com/EasyRPG/Player/blob/master/src/decoder_libsndfile.cpp#L28

**Comment by evpobr** [comment]
Changing `sf_open()` behavour requires major version bump (v2.0.0). We are not yet ready for this.

But we can add `sf_utf8_open()` function with minor bump (v1.1.0). @erikd , what do you think?

**Comment by brianlheim** [comment]
Sorry, why does it require a bump? It doesn't break any existing functionality, just expands what filenames are accepted, unless I'm misunderstanding something. Regardless, all I'm really after is a UTF-8 const char * function, so `sf_utf8_open` would also be fine.

**Comment by evpobr** [comment]
> It doesn't break any existing functionality, just expands what filenames are accepted, unless I'm misunderstanding something.

No. Existings apps will still use ANSI paths but UTF-8 is compatible with 7-bit US-ANSII only , lower 127 codes of any 8-bit encoding table (latin letters, numbers, punctuation etc). But what is path is in Windows-1251 and uses higher codes (Cyrillic letters for example)?  Encoding to UTF-8 will fail, encoded path will contain garbage.

**Comment by brianlheim** [comment]
Oh right! For some reason I thought the Winapi OpenFile function worked with the ASCII set only, and that there was a separate function for other codepages. I see now that it expects ANSI. My mistake!

**Comment by erikd** [comment]
> But we can add sf_utf8_open() function with minor bump (v1.1.0). @erikd , what do you think?

@evpobr You mean add this as a Windows only complement to `sf_wchar_open()`?

**Comment by evpobr** [comment]
For all plaftorms as uniform way to use unicode.

**Comment by erikd** [comment]
But Linux and Mac already handle UTF-8 perfectly with `sf_open()`. Why then do Windows and Mac need `sf_utf8_open()` ?

If we need a version bump so that `sf_open()` switches to support UTF-8 on Windows, then that may be what we need to do.

**Comment by evpobr** [comment]
> But Linux and Mac already handle UTF-8 perfectly with sf_open(). Why then do Linux and Mac need sf_utf8_open() ?

To have one open function across all platforms if you are UTF-8 aware.

But UTF-8 Everywhere is utopia without hacks:

* You still need to interact with `wchar_t` using WinAPI
* You still need to interact with ANSI-chars using C standard library under Windows
* Linux is NOT garanteed 100% to be UTF-8 aware, it is "silent" standard

Whatever you do, you need `#ifdef`'s anyway.

So let's leave it as it is. You can open files with unicode paths with SndFile on every platform, thanks to `sf_open_fd()`, `sf_wchar_open()` and `sf_open_virtual()`.

**Comment by brianlheim** [comment]
> Whatever you do, you need #ifdef's anyway.

But, by not providing a UTF-8 function in your library, you force all users of your library to write these ifdefs again and again. I am asking you to write them once instead. You could be eliminating a lot of duplicate code and margins for client error if you added this capability.

> Linux is NOT garanteed 100% to be UTF-8 aware, it is "silent" standard

If you are really concerned about this, you can use the same approach described in the last row of this table: https://www.boost.org/doc/libs/1_63_0/libs/filesystem/doc/reference.html#Operating-system-examples. This is much more compliant to actual standards than what I'm suggesting, which would get you 99% of the way there anyway. The more rigorous solution is basically that on non-macOS POSIX systems you would use the locale given by `setlocale`; there is also another check to do on Windows. Or, you can just say that on Windows this function is expected to take UTF-8, and on other systems is expected to be in the native path locale, which is almost always UTF-8.

> You can open files with unicode paths with SndFile on every platform, thanks to sf_open_fd(), sf_wchar_open() and sf_open_virtual().

It is impossible to use these functions dependably in a cross-platform manner, which is why I opened this ticket. `sf_open_virtual` is not a good alternative, as it also demands extra client code for what should be a simple task in this library. What is needed is a _single_ function that can be used on any platform to open a sound file, without extra wrapping code on the part of the client.

**Comment by evpobr** [comment]
Another idea - new `SFM_UTF8` flag for `sf_open`. It can be used on Windows and ignored on all other platforms.

**Comment by evpobr** [comment]
@erikd, cannot start without your opinion - `sf_utf8_open` function or `SFM_UTF8` flags for `sf_open()`?
